File: solutions/python/src/shared/outbox_relay.py
"""R8: the ONLY component that moves events from the Outbox to the bus.

Runs after the unit of work committed (Day 6: a polling loop; here: called explicitly).
If publish raises we do not mark the row, so it is retried -> at-least-once.
"""
from shared.event_bus import InMemoryEventBus
from shared.outbox import Outbox


class OutboxRelay:
    def __init__(self, outbox: Outbox, bus: InMemoryEventBus) -> None:
        self._outbox = outbox
        self._bus = bus

    def publish_pending(self) -> int:
        events = self._outbox.pending()
        for event in events:
            self._bus.publish(event)
            self._outbox.mark_published(event.event_id)
        return len(events)

File: solutions/node/src/shared/OutboxRelay.ts
import type { Outbox } from './Outbox.ts';
import type { InMemoryEventBus } from './EventBus.ts';

// R8: the ONLY component that moves events from the Outbox to the bus.
// Runs after the unit of work committed (Day 6: a polling loop / cron; here: called explicitly).
// If publish throws we do not mark the row, so it is retried -> at-least-once.
export class OutboxRelay {
  private readonly outbox: Outbox;
  private readonly bus: InMemoryEventBus;

  constructor(outbox: Outbox, bus: InMemoryEventBus) {
    this.outbox = outbox;
    this.bus = bus;
  }

  async publishPending(): Promise<number> {
    const events = await this.outbox.pending();
    for (const event of events) {
      await this.bus.publish(event);
      await this.outbox.markPublished(event.eventId);
    }
    return events.length;
  }
}
